### DISCUSSION TITLE
Compose on alt keys with Kanata

### DISCUSSION THREAD

**Discussion by magthe** [YaLTeR/niri#878] [General]
I have a setup where I use [kanata](https://github.com/jtroo/kanata) to remap some keys, including to get compose when tapping either alt key.  I've used this setup successfully under both Sway and Hyprland, but with niri the alt keys always emits alt.

The other parts of the configuration work as expected, e.g. esc on tapping CAPS, and arrows on hjkl while holding CAPS.

Is there something niri does that would explain this, and is it possible to configure it to not do it?

My full config for kanata:

```
(defcfg
  linux-dev /dev/input/by-path/platform-i8042-serio-0-event-kbd
  process-unmapped-keys yes
)

(defsrc
  esc
                                                                 p     lbrc
  caps                            h     j       k        l       scln  apo
  lmet  lalt                   ralt   rctrl    lft
)

(deflayer default
  caps
                                                                 _     _
  @esc                            _     _       _        _       _     _
  _     @comp                  @comp  lmet     rctrl
)

(deflayer navigation
  _
                                                                 pgup  home
  _                               left  down    up       rght    pgdn  end
  _      _                     _      _        _
)

(defalias
  esc (tap-hold-press 150 150 esc (layer-toggle navigation))
  comp (tap-hold-press 100 100 ralt lalt)
  )
```

**Comment by YaLTeR** [comment]
Not sure how kanata works, by the looks of it somewhere on the event device level so it should work? Especially since other overrides work fine. Niri doesn't do anything special with alt.

Sidenote, there might be an xkb option to put compose on alt keys.

**Comment by magthe** [comment]
Yes, there's an xkb option to do that, but then I lose the ability to use alt for anything else. The tap-hold feature is what I _really_ like with kanata.

**Comment by magthe** [comment]
Silly me!

It turns out it was my config that was at fault. After adding

```
input {
    keyboard {
        xkb {
            layout "us(altgr-intl)"
            options "compose:ralt"
        }
    }
    ...
}
```

it now works as expected.
